crates/sargon/src/system/sargon_os/pre_authorization/sargon_os_pre_authorization_status.rs

```diff
@@ -0,0 +1,392 @@
+use crate::prelude::*;
+use std::time::Duration;
+
+// ==================
+// Poll PreAuthorization Status (Public)
+// ==================
+impl SargonOS {
+    /// Polls the state of a PreAuthorization until we can determine its `TransactionStatus`.
+    pub async fn poll_pre_authorization_status(
+        &self,
+        intent_hash: SubintentHash,
+        expiration: Option<DappToWalletInteractionSubintentExpiration>,
+    ) -> Result<TransactionStatus> {
+        let (status, _) = self
+            .poll_subintent_status_with_delays(intent_hash, expiration)
+            .await?;
+
+        Ok(status)
+    }
+}
+
+// ==================
+// Poll PreAuthorization Status (Internal)
+// ==================
+impl SargonOS {
+    /// Polls the state of a Subintent until it is submitted
+    ///
+    /// It returns the `TransactionIntentHash`, but also the list of delays between each poll.
+    async fn poll_subintent_status_with_delays(
+        &self,
+        intent_hash: SubintentHash,
+        expiration: Option<DappToWalletInteractionSubintentExpiration>,
+    ) -> Result<(TransactionStatus, Vec<u64>)> {
+        let network_id = self.current_network_id()?;
+        let gateway_client = GatewayClient::new(
+            self.clients.http_client.driver.clone(),
+            network_id,
+        );
+
+        let expiration_timestamp = self.expiration_timestamp(expiration);
+
+        let mut delays: Vec<u64> = vec![];
+
+        // The delay increment is set to 1 second in production, but 1 millisecond in tests.
+        // This will make the tests run with almost no delay, while the production code will have a 2s delay after first call,
+        // a 3s delay after second call, 4s after third and so on.
+        #[cfg(test)]
+        const DELAY_INCREMENT: u64 = 1;
+        #[cfg(not(test))]
+        const DELAY_INCREMENT: u64 = 1000;
+
+        let mut delay_duration = DELAY_INCREMENT;
+
+        loop {
+            // Check if the subinent hasn't expired already
+            if Timestamp::now_utc() > expiration_timestamp {
+                return Err(CommonError::ExpiredSubintent);
+            }
+
+            // Increase delay by 1 second on subsequent calls
+            delay_duration += DELAY_INCREMENT;
+            let sleep_duration = Duration::from_millis(delay_duration);
```

Comment:
I am not entirely sure this logic is correct for subintent polling. We should likely take into account the expiration duration and adjust the delay accordingly. If say the dApp specified the expiration to be 25 seconds, and it submits the parent transaction at second 21 and it is committed fully to the network before expiration, say second 24, it seems that this polling will consider the subintent as expired, because we will have the following times for checking:
- t1 = 0
- t2 = 2
- t3 = 5
- t4 = 9
- t5 = 14
- t6 = 20
- t7 = 27

And the checking at t7 will hit the expiration check above while the subintent was actually submitted. Let me know if I reading this code wrongly :).

Reply:
good catch! Added this consideration to the expiration check